proposals/0371-isolated-synchronous-deinit.md

```diff
@@ -117,94 +124,161 @@ internal func _deinitOnExecutor(_ object: __owned AnyObject,
 
-Non-deallocating `deinit` is not affected by this proposal.
+If `deinit` is isolated, code that normally is emitted into `__deallocating_init` gets emitted into a new entity (`__isolated_deallocating_init`), and `__deallocating_init` is emitted as a thunk that reads the executor (from `self` for actors and from the global actor for GAITs) and calls corresponding runtime function passing `self`, `__isolated_deallocating_init` and the desired executor.
+
+Destroying `deinit` is not affected by this proposal.
 
 ### Rules for computing isolation
 
-Isolation of `deinit` comes with runtime and code size cost. Types that don't perform custom actions in `deinit` and only need to release references don't need isolated `deinit`. Releasing child objects can be done from any thread. If those objects are concerned about isolation, they should adopt isolation themselves.
+For backwards compatibility, isolation of the class is not propagated to the synchronous `deinit` by default:
 
 ```swift
 @MainActor
 class Foo {
-    let bar: Bar
-    
-    // No isolated deinit generated.
-    // Reference to Bar can be released from any thread.
-    // Class Bar is responsible for correctly isolating its own deinit.
+  deinit {} // not isolated
 }
+```
 
-actor MyActor {
-    let bar: Bar
-    
-    // Similar
+To opt-in into isolation propagation, proposal allows `isolated` attribute to be applied to `deinit` declaration. This is an existing attribute, which currently can be applied only to function arguments, but has a different meaning there. When applied to a function argument it indicates a parameter which contains an instance of the actor which function should be isolated to.
+
+```swift
+@MainActor
+class Foo {
+  isolated deinit {} // Isolated on MainActor.shared
+}
+
+actor Bar {
+  isolated deinit {} // Isolated on self
 }
 ```
 
-If there is an explicit `deinit`, then isolation is computed following usual rules for isolation of class instance members as defined by [SE-0313](https://github.com/gottesmm/swift-evolution/blob/move-function-pitch-v1/proposals/0313-actor-isolation-control.md) and [SE-0316](https://github.com/apple/swift-evolution/blob/main/proposals/0316-global-actors.md). It takes into account isolation attributes on the `deinit` itself, isolation of the `deinit` in the superclass, and isolation attributes on the containing class. If deinit belongs to an actor or GAIT, but isolation of the `deinit` is undesired, it can be suppressed using `nonisolated` attribute:
+It is an error to use `isolated` attribute on a `deinit`, if containing class has no isolation.
 
 ```swift
-@MainActor
 class Foo {
-    deinit {} // Isolated on MainActor
+  // error: deinit is marked isolated, but containing class Foo is not isolated to an actor
+  isolated deinit {}
 }
+```
 
-@FirstActor
-class Bar {
-    @SecondActor
-    deinit {} // Isolated on SecondActor
+If containing class is not isolated, it is still possible to use global actor attribute on `deinit`.
```

Comment:
```suggestion
If the containing type is not isolated, it is still possible to use a global actor attribute on `deinit`.
```

Reply:
I don't want to use `type` to avoid getting false idea that this feature can be used with move-only value types.